File: diff/comparison/chunk_optimizer.go
// Copyright 2000-2021 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
// Ported from intellij-community to Go.

package comparison

import (
	"github.com/byron1st/file-diff/diff/myers"
	"github.com/byron1st/file-diff/diff/util"
)

// optimizeLineChunks adjusts chunk boundaries to prefer empty/unimportant lines
// as natural boundaries for insertions and deletions.
func optimizeLineChunks(lines1, lines2 []*Line, iterable FairDiffIterable) FairDiffIterable {
	ranges := buildOptimizedRanges(lines1, lines2, iterable)
	return Fair(CreateUnchanged(ranges, len(lines1), len(lines2)))
}

func buildOptimizedRanges(lines1, lines2 []*Line, iterable FairDiffIterable) []util.Range {
	var ranges []util.Range

	for _, r := range iterable.Unchanged() {
		ranges = append(ranges, r)
		processLastRanges(lines1, lines2, &ranges)
	}
	return ranges
}

func processLastRanges(lines1, lines2 []*Line, ranges *[]util.Range) {
	if len(*ranges) < 2 {
		return
	}

	r1 := (*ranges)[len(*ranges)-2]
	r2 := (*ranges)[len(*ranges)-1]

	if r1.End1 != r2.Start1 && r1.End2 != r2.Start2 {
		return
	}

	count1 := r1.End1 - r1.Start1
	count2 := r2.End1 - r2.Start1

	eqFwd := expandForwardLines(lines1, lines2, r1.End1, r1.End2, r1.End1+count2, r1.End2+count2)
	eqBwd := expandBackwardLines(lines1, lines2, r2.Start1-count1, r2.Start2-count1, r2.Start1, r2.Start2)

	if eqFwd == 0 && eqBwd == 0 {
		return
	}

	// merge left
	if eqFwd == count2 {
		*ranges = (*ranges)[:len(*ranges)-2]
		*ranges = append(*ranges, util.NewRange(r1.Start1, r1.End1+count2, r1.Start2, r1.End2+count2))
		processLastRanges(lines1, lines2, ranges)
		return
	}

	// merge right
	if eqBwd == count1 {
		*ranges = (*ranges)[:len(*ranges)-2]
		*ranges = append(*ranges, util.NewRange(r2.Start1-count1, r2.End1, r2.Start2-count1, r2.End2))
		processLastRanges(lines1, lines2, ranges)
		return
	}

	touchSideIsLeft := r1.End1 == r2.Start1
	shift := getLineShift(lines1, lines2, touchSideIsLeft, eqFwd, eqBwd, r1, r2)
	if shift != 0 {
		*ranges = (*ranges)[:len(*ranges)-2]
		*ranges = append(*ranges,
			util.NewRange(r1.Start1, r1.End1+shift, r1.Start2, r1.End2+shift),
			util.NewRange(r2.Start1+shift, r2.End1, r2.Start2+shift, r2.End2),
		)
	}
}

func getLineShift(lines1, lines2 []*Line, touchSideIsLeft bool, eqFwd, eqBwd int, r1, r2 util.Range) int {
	threshold := myers.UnimportantLineCharCount()

	var touchLines []*Line
	var touchStart int
	if touchSideIsLeft {
		touchLines = lines1
		touchStart = r2.Start1
	} else {
		touchLines = lines2
		touchStart = r2.Start2
	}

	// Try unchanged boundary shift
	if s := findBoundaryShift(touchLines, touchStart, eqFwd, eqBwd, 0); s != nil {
		return *s
	}

	// Try changed boundary shift
	var nonTouchLines []*Line
	var changeStart, changeEnd int
	if touchSideIsLeft {
		nonTouchLines = lines2
		changeStart = r1.End2
		changeEnd = r2.Start2
	} else {
		nonTouchLines = lines1
		changeStart = r1.End1
		changeEnd = r2.Start1
	}
	if s := findBoundaryShift(nonTouchLines, changeStart, eqFwd, eqBwd, 0); s != nil {
		return *s
	}

	// Try with threshold
	if s := findBoundaryShift(touchLines, touchStart, eqFwd, eqBwd, threshold); s != nil {
		return *s
	}
	if s := findBoundaryShift(nonTouchLines, changeStart, eqFwd, eqBwd, threshold); s != nil {
		_ = changeEnd // suppress unused
		return *s
	}

	return 0
}

func findBoundaryShift(lines []*Line, offset, eqFwd, eqBwd, threshold int) *int {
	fwd := findNextUnimportant(lines, offset, eqFwd+1, threshold)
	bwd := findPrevUnimportant(lines, offset-1, eqBwd+1, threshold)

	if fwd == -1 && bwd == -1 {
		return nil
	}
	if fwd == 0 || bwd == 0 {
		zero := 0
		return &zero
	}
	if fwd != -1 {
		return &fwd
	}
	neg := -bwd
	return &neg
}

func findNextUnimportant(lines []*Line, offset, count, threshold int) int {
	for i := range count {
		if lines[offset+i].nonSpaceChars <= threshold {
			return i
		}
	}
	return -1
}

func findPrevUnimportant(lines []*Line, offset, count, threshold int) int {
	for i := range count {
		if lines[offset-i].nonSpaceChars <= threshold {
			return i
		}
	}
	return -1
}

func expandForwardLines(lines1, lines2 []*Line, start1, start2, end1, end2 int) int {
	s1, s2 := start1, start2
	for s1 < end1 && s2 < end2 && lines1[s1].equals(lines2[s2]) {
		s1++
		s2++
	}
	return s1 - start1
}

func expandBackwardLines(lines1, lines2 []*Line, start1, start2, end1, end2 int) int {
	e1, e2 := end1, end2
	for start1 < e1 && start2 < e2 && lines1[e1-1].equals(lines2[e2-1]) {
		e1--
		e2--
	}
	return end1 - e1
}

File: swift/Sources/FileDiff/Internal/ChunkOptimizer.swift
// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
// Ported from intellij-community to Swift (via the Go port in this repository).

/// Adjusts chunk boundaries so that insertions and deletions are preferably
/// aligned with empty or otherwise "unimportant" lines.
func optimizeLineChunks(_ lines1: [Line], _ lines2: [Line], _ iterable: FairDiffIterable) -> FairDiffIterable {
    var ranges: [DiffRange] = []
    for range in iterable.unchanged {
        ranges.append(range)
        processLastRanges(lines1: lines1, lines2: lines2, ranges: &ranges)
    }
    return createUnchanged(ranges, length1: lines1.count, length2: lines2.count)
}

// swiftlint:disable function_body_length
private func processLastRanges(lines1: [Line], lines2: [Line], ranges: inout [DiffRange]) {
    // swiftlint:enable function_body_length
    if ranges.count < 2 {
        return
    }
    let r1 = ranges[ranges.count - 2]
    let r2 = ranges[ranges.count - 1]

    if r1.left.upperBound != r2.left.lowerBound, r1.right.upperBound != r2.right.lowerBound {
        return
    }

    let count1 = r1.left.count
    let count2 = r2.left.count

    let eqFwd = expandForwardLines(
        lines1,
        lines2,
        r1.left.upperBound,
        r1.right.upperBound,
        r1.left.upperBound + count2,
        r1.right.upperBound + count2
    )
    let eqBwd = expandBackwardLines(
        lines1,
        lines2,
        r2.left.lowerBound - count1,
        r2.right.lowerBound - count1,
        r2.left.lowerBound,
        r2.right.lowerBound
    )

    if eqFwd == 0, eqBwd == 0 {
        return
    }

    if eqFwd == count2 {
        ranges.removeLast(2)
        ranges.append(DiffRange(
            start1: r1.left.lowerBound, end1: r1.left.upperBound + count2,
            start2: r1.right.lowerBound, end2: r1.right.upperBound + count2
        ))
        processLastRanges(lines1: lines1, lines2: lines2, ranges: &ranges)
        return
    }
    if eqBwd == count1 {
        ranges.removeLast(2)
        ranges.append(DiffRange(
            start1: r2.left.lowerBound - count1, end1: r2.left.upperBound,
            start2: r2.right.lowerBound - count1, end2: r2.right.upperBound
        ))
        processLastRanges(lines1: lines1, lines2: lines2, ranges: &ranges)
        return
    }

    let touchSideIsLeft = r1.left.upperBound == r2.left.lowerBound
    let shift = getLineShift(
        lines1: lines1,
        lines2: lines2,
        touchSideIsLeft: touchSideIsLeft,
        eqFwd: eqFwd,
        eqBwd: eqBwd,
        r1: r1,
        r2: r2
    )
    if shift != 0 {
        ranges.removeLast(2)
        ranges.append(DiffRange(
            start1: r1.left.lowerBound, end1: r1.left.upperBound + shift,
            start2: r1.right.lowerBound, end2: r1.right.upperBound + shift
        ))
        ranges.append(DiffRange(
            start1: r2.left.lowerBound + shift, end1: r2.left.upperBound,
            start2: r2.right.lowerBound + shift, end2: r2.right.upperBound
        ))
    }
}

// swiftlint:disable function_parameter_count
private func getLineShift(
    lines1: [Line],
    lines2: [Line],
    touchSideIsLeft: Bool,
    eqFwd: Int,
    eqBwd: Int,
    r1: DiffRange,
    r2: DiffRange
) -> Int {
    // swiftlint:enable function_parameter_count
    let threshold = unimportantLineCharCount

    let touchLines: [Line]
    let touchStart: Int
    if touchSideIsLeft {
        touchLines = lines1
        touchStart = r2.left.lowerBound
    } else {
        touchLines = lines2
        touchStart = r2.right.lowerBound
    }

    if let shift = findBoundaryShift(touchLines, touchStart, eqFwd, eqBwd, 0) {
        return shift
    }

    let nonTouchLines: [Line]
    let changeStart: Int
    if touchSideIsLeft {
        nonTouchLines = lines2
        changeStart = r1.right.upperBound
    } else {
        nonTouchLines = lines1
        changeStart = r1.left.upperBound
    }
    if let shift = findBoundaryShift(nonTouchLines, changeStart, eqFwd, eqBwd, 0) {
        return shift
    }
    if let shift = findBoundaryShift(touchLines, touchStart, eqFwd, eqBwd, threshold) {
        return shift
    }
    if let shift = findBoundaryShift(nonTouchLines, changeStart, eqFwd, eqBwd, threshold) {
        return shift
    }
    return 0
}

private func findBoundaryShift(_ lines: [Line], _ offset: Int, _ eqFwd: Int, _ eqBwd: Int, _ threshold: Int) -> Int? {
    let fwd = findNextUnimportant(lines, offset, eqFwd + 1, threshold)
    let bwd = findPrevUnimportant(lines, offset - 1, eqBwd + 1, threshold)

    if fwd == -1, bwd == -1 {
        return nil
    }
    if fwd == 0 || bwd == 0 {
        return 0
    }
    if fwd != -1 {
        return fwd
    }
    return -bwd
}

private func findNextUnimportant(_ lines: [Line], _ offset: Int, _ count: Int, _ threshold: Int) -> Int {
    for i in 0..<count where lines[offset + i].nonSpaceChars <= threshold {
        return i
    }
    return -1
}

private func findPrevUnimportant(_ lines: [Line], _ offset: Int, _ count: Int, _ threshold: Int) -> Int {
    for i in 0..<count where lines[offset - i].nonSpaceChars <= threshold {
        return i
    }
    return -1
}

// swiftlint:disable function_parameter_count
private func expandForwardLines(
    _ lines1: [Line],
    _ lines2: [Line],
    _ start1: Int,
    _ start2: Int,
    _ end1: Int,
    _ end2: Int
) -> Int {
    // swiftlint:enable function_parameter_count
    var s1 = start1
    var s2 = start2
    while s1 < end1, s2 < end2, lines1[s1].equals(lines2[s2]) {
        s1 += 1
        s2 += 1
    }
    return s1 - start1
}

// swiftlint:disable function_parameter_count
private func expandBackwardLines(
    _ lines1: [Line],
    _ lines2: [Line],
    _ start1: Int,
    _ start2: Int,
    _ end1: Int,
    _ end2: Int
) -> Int {
    // swiftlint:enable function_parameter_count
    var e1 = end1
    var e2 = end2
    while start1 < e1, start2 < e2, lines1[e1 - 1].equals(lines2[e2 - 1]) {
        e1 -= 1
        e2 -= 1
    }
    return end1 - e1
}
